include/openssl/core.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_CORE_H
+# define OSSL_CORE_H
+
+# include <stddef.h>
+
+# ifdef __cplusplus
+extern "C" {
+# endif
+
+/*-
+ * Base types
+ * ----------
+ *
+ * These are the types that the OpenSSL core and providers have in common
+ * to communicate data between them.
+ */
+
+/*
+ * Dispatch table element.  function_id numbers are defined further down,
+ * see macros with '_FUNC' in their names.
+ *
+ * An array of these is always terminated by function_id == 0
+ */
+struct ossl_dispatch_st {
+    int function_id;
+    void (*function)(void);
+};
+
+/*
+ * Other items, essentially an int<->pointer map element.
+ * We could have used this as the dispatch table element as well, if it
+ * wasn't for certain compilers that warn with you store a function pointer
+ * in a non-function pointer variable.
+ *
+ * This is used whenever we need to pass things like a table of error reason
+ * codes <-> reason string maps, parameter name <-> parameter type maps, ...
+ *
+ * Usage determines which field works as key if any, rather than field order.
+ *
+ * An array of these is always terminated by id == 0 && ptr == NULL
+ */
+struct ossl_item_st {
+    int id;
+    void *ptr;
+};
```

Comment:
This is missing in the design document, I'll submit a proposal for it in the next few days.